etcd/v3/kv_etcd.go

```diff
@@ -1497,53 +1497,84 @@ func (et *etcdKV) RemoveMember(
 }
 
 func (et *etcdKV) ListMembers() (map[string]*kvdb.MemberInfo, error) {
+	var (
+		fnMemberStatus = func(cliURL string) (*kvdb.MemberInfo, uint64, error) {
+			if cliURL == "" {
+				return nil, 0, fmt.Errorf("Must provide client URL")
+			}
+			// Use the context with no leader requirement as we might be hitting
+			// an endpoint which is down
+			ctx, cancel := et.MaintenanceContext()
+			endpointStatus, err := et.maintenanceClient.Status(ctx, cliURL)
+			cancel()
+
+			if err != nil {
+				return nil, 0, err
+			}
+
+			mid := endpointStatus.Header.MemberId
+			return &kvdb.MemberInfo{
+				// PeerUrls:   .. must fill later
+				ClientUrls: []string{cliURL},
+				Leader:     endpointStatus.Leader == mid,
+				DbSize:     endpointStatus.DbSize,
+				IsHealthy:  true,
+				ID:         strconv.FormatUint(mid, 16),
+			}, mid, nil
+		}
+	)
 	ctx, cancel := et.MaintenanceContextWithLeader()
 	memberListResponse, err := et.kvClient.MemberList(ctx)
 	cancel()
+
 	if err != nil {
 		return nil, err
 	}
-	resp := make(map[string]*kvdb.MemberInfo)
+
+	membersMap := make(map[uint64]*kvdb.MemberInfo)
 	mLock.Lock()
 	defer mLock.Unlock()
+
+	// Get status from Endpoints
+	for _, ep := range et.GetEndpoints() {
+		mi, mid, err := fnMemberStatus(ep)
+		if err != nil || mi == nil {
+			logrus.WithError(err).Warnf("kvClient.Status(%s) returned error", ep)
+			continue
+		}
+		membersMap[mid] = mi
+	}
+
+	// Get status from MemberList() nodes, that were missing from GetEndpoints() list
 	for _, member := range memberListResponse.Members {
-		var (
-			leader     bool
-			dbSize     int64
-			isHealthy  bool
-			clientURLs []string
-		)
-		// etcd versions < v3.2.15 will return empty ClientURLs if
-		// the node is unhealthy. For versions >= v3.2.15 they populate
-		// ClientURLs but return an error status
-		if len(member.ClientURLs) != 0 {
-			// Use the context with no leader requirement as we might be hitting
-			// an endpoint which is down
-			ctx, cancel := et.MaintenanceContext()
-			endpointStatus, err := et.maintenanceClient.Status(
-				ctx,
-				member.ClientURLs[0],
-			)
-			cancel()
-			if err == nil {
-				if member.ID == endpointStatus.Leader {
-					leader = true
+		for _, cu := range member.ClientURLs {
+			if _, has := membersMap[member.ID]; !has {
+
+				mi, mid, err := fnMemberStatus(cu)
+				if err != nil || mi == nil {
+					logrus.WithError(err).Warnf("kvClient.Status(%s) returned error", cu)
+					continue
 				}
-				dbSize = endpointStatus.DbSize
-				isHealthy = true
-				// Only set the urls if status is healthy
-				clientURLs = member.ClientURLs
+				membersMap[mid] = mi
```

Comment:
could do a break her and avoid the loop

Reply:
Don't think breaking out the loop will change anything here --
* the loops will go through _all_ `member.ClientURLs` elements, but ...
* the first client URL that worked, will set up its `membersMap[member.ID]` element in the map (line 1558), and
* all consecutive ClientURLs for the same client will be skipped  (as `membersMap[member.ID]` element will be present -- line 1551)
* it probably wasn't very obvious that `member.ID == mid`  (they are equivalent)